source/dofs/dof_tools.cc

```diff
@@ -1197,39 +1197,11 @@ namespace DoFTools
   extract_locally_relevant_dofs(const DoFHandler<dim, spacedim> &dof_handler,
                                 IndexSet &                       dof_set)
   {
-    // collect all the locally owned dofs
-    dof_set = dof_handler.locally_owned_dofs();
-
-    // now add the DoF on the adjacent ghost cells to the IndexSet
-
-    // Note: For certain meshes (in particular in 3D and with many
-    // processors), it is really necessary to cache intermediate data. After
-    // trying several objects such as std::set, a vector that is always kept
-    // sorted, and a vector that is initially unsorted and sorted once at the
-    // end, the latter has been identified to provide the best performance.
-    // Martin Kronbichler
-    std::vector<types::global_dof_index> dof_indices;
-    std::vector<types::global_dof_index> dofs_on_ghosts;
+    const auto &partitioner =
+      *dof_handler.locally_relevant_dofs_partitioner().lock();
 
-    typename DoFHandler<dim, spacedim>::active_cell_iterator
-      cell = dof_handler.begin_active(),
-      endc = dof_handler.end();
-    for (; cell != endc; ++cell)
-      if (cell->is_ghost())
-        {
-          dof_indices.resize(cell->get_fe().n_dofs_per_cell());
-          cell->get_dof_indices(dof_indices);
-          for (const auto dof_index : dof_indices)
-            if (!dof_set.is_element(dof_index))
-              dofs_on_ghosts.push_back(dof_index);
-        }
-
-    // sort, compress out duplicates, fill into index set
-    std::sort(dofs_on_ghosts.begin(), dofs_on_ghosts.end());
-    dof_set.add_indices(dofs_on_ghosts.begin(),
-                        std::unique(dofs_on_ghosts.begin(),
-                                    dofs_on_ghosts.end()));
-    dof_set.compress();
+    dof_set = partitioner.locally_owned_range();
+    dof_set.add_indices(partitioner.ghost_indices());
```

Comment:
That's an expensive operation. Why not just store the whole locally relevant index set in `DoFHandler` as well? I believe that we have wanted to do that for a long time already.

Reply:
I would not think this is too expensive, at least if we did the implementation right. The locally owned range is typically only one or at most a few index ranges, so that should be pretty fast. And if it is not, we should make it fast. The question is then whether we want to keep two data structures with very similar content (partitioner and index set). But in any case, we need to think of how to handle the case with non-contiguous locally owned indices, which is currently not within the capabilities of the `Partitioner` class.

Reply:
> But in any case, we need to think of how to handle the case with non-contiguous locally owned indices, which is currently not within the capabilities of the Partitioner class.

I still need to understand why the test fail now. In anyway, we need partitioners (to set up vectors). Or is there an intermediate stage that causes the break? I will investigate in the next days.